glommio/src/channels/spsc_queue.rs

```diff
@@ -115,7 +129,8 @@ impl<T: Copy> Buffer<T> {
             }
         }
 
-        let v = self.buffer_storage[current_head % self.ccache.capacity].get();
+        let cell = self.buffer_storage[current_head % self.ccache.capacity].as_ptr();
+        let v = unsafe { ptr::read(cell) };
```

Comment:
Hi @thirstycrow 

This is a great change, as I mentioned before. But because it is very impactful I'd like to see it in a separate PR.
We should also make @bryandmc aware since he was working on a proposed implementation of that too.

Would you be able to clean this up and send just this part as a separate PR?
Ack the direction!

However I am concerned about `ptr::read` and `ptr::write` here. I just double checked the docs and they say, respectively:

```
Reads the value from src without moving it. This leaves the memory in src unchanged.
```

```
Overwrites a memory location with the given value without reading or dropping the old value.
```

So does that perform a copy? Bryan's approach was without this copy, passing ownership of the object into the channel (although I have not seen the code). For a high throughput channel, that is preferable

Reply:
Hey I am on vacation so I haven't had a chance to catch up on all this but I will mention that ptr read and write could be used in the right circumstances to help accomplish the goal. But I'd have to see exactly how it's being used.  Mostly I just wanted to mention that I'm away and won't be able to finish up that code or look into this stuff till Tuesday.

Reply:
hope you had fun!

Reply:
TLDR: Change Cell -> UnsafeCell in the SPSC queue, or enum with destructured data from heap allocated types, and directly copying for copy types. 

Ok so I've gone done a few different paths to see what works best.. First, I went down the path where I left alone the actual queue and it's structure (namely, the copy bound) and managed to make it _more_ ergonomic.. But it has drawbacks. The other method, is replacing the `Cell` with an `UnsafeCell` s/t we can remove the `T: Copy` bound entirely, while still doing internal mutability in a sane way. The drawbacks of the `UnsafeCell` method is that you can have very different sized entries, based on what the user decides to use. But, on the other hand, this is what the standard library does. It gives you that option to put a large struct into one of these data structures, if you wanted to. The other method is pretty simple in theory, but heavy with lots of glue code to make it ergonomic. What the enum method does is create a structure like this: 

```rust
pub enum Shared<T: Send, V: Send + Copy> {
    /// Boxed data
    Box(NonNull<T>),
    /// Strings
    String(NonNull<u8>, usize, usize),
    /// Vectors
    Vec(NonNull<T>, usize, usize),
    /// Copy types that are ok to actually just copy
    Copy(V),
}
```
That includes many of the most common types of heap allocated structures and any data needed to deconstruct/reconstruct, as well as a variant for any type that is Copy, to be stored directly, as is. With a little bit of trait and impl magic you can get to the point where the user's code looks like:

```rust
try_send(String::from("some string data here.."));
try_send(vec!["a".to_string(), "b".to_string()]);
try_send(Box::new(String::from("something here..")));
try_send(100usize);
```
So it can probably be made relatively ergonomic.. That's good. The try_send or any method that needs to take these values can just use a generic or the impl Trait form. (Note: ConvertInto is the trait I had to make to get around the issue of overlapping impls. Basically, can't be don't with normal From/Into.. sadly..)

For example:
```rust
fn try_send<T, V>(i: impl ConvertInto<Shared<T, V>>)
where
    T: Send,
    V: Send + Copy,
{
    // ... do stuff ...
}
```
The UnsafeCell method is actually probably gonna end up being a lot simpler.. All an UnsafeCell has from an API standpoint is `UnsafeCell::get(&self) -> *mut T` which will give you a mutable pointer to it's internals. Then we can safely just move values into that position, and get them back out with `ptr::read`.  The truth is there will probably always need to be ptr::read somewhere in this process.. In any of these options you need the data contained in a Vec struct (for example) to be able to produce the value on the other side.. whether that's with the UnsafeCell or with my custom enum.. either way we probably won't be able to just move out of the Buffer once it gets put there. At least I don't see a way. To be fair, this is how most standard library structures are implemented in rust. For example `Vec::pop(&mut self) -> Option<T>`:

```rust
pub fn pop(&mut self) -> Option<T> {
    if self.len == 0 {
        None
    } else {
        unsafe {
            self.len -= 1;
            Some(ptr::read(self.as_ptr().add(self.len())))
        }
    }
}
```
The only way (that I know) to produce an owned value from a pointer in rust, when the `T` isn't Copy is ptr::read. 

So I'm going to try and finish implementing the UnsafeCell version and see how that goes, first. I've begun implementing the enum version as well, but I'm gonna put that one on the backburner for now because of the increased simplicity of the UnsafeCell version.

NOTE: My original idea _seemed_ to have less copies, but it also proved to be unsound and had ergonomic issues that couldn't be resolved. After studying some other containers in std and in the community, this seemed like the way to go.